src/main/java/com/datatorrent/netlet/AbstractClient.java

```diff
@@ -45,7 +91,11 @@
   protected CircularBuffer<Slice> sendBuffer4Offers, sendBuffer4Polls;
   protected final ByteBuffer writeBuffer;
   protected boolean write = true;
-  protected SelectionKey key;
+  /*
+   * access to the key is not thread safe. It is read/write on the default event loop and read only on other threads,
+   * so other threads may get stale value.
+   */
+  protected SelectionKey key = invalidSelectionKey;
```

Comment:
I was considering to change it to volatile or use Unsafe. We may need to revisit how `key` is used.